### PR TITLE
Digest class moved to util directory.

### PR DESCRIPTION
**Pull request by pkthapa**
**Changes mentioned here**:
- `STDigest` class removed. `Digest` class used in place of `STDigest`. Both these classes use Crypto++ library for hashing.
- Move `Digest` class from `bftengine` to `util` directory.

Note: These changes are done as a part of aligning cryptographic algorithms task.

### CODE REVIEW

**Review by arc-vmware — COMMENTED** [review]
@pkthapa, thanks for doing the changes. Can you please give a maestro run.

File: util/src/Digest.cpp [hunk 1]
```diff
@@ -98,8 +98,9 @@ Digest::Digest(char* buf, size_t len) { DigestUtil::compute(buf, len, (char*)d,
 
 std::string Digest::toString() const {
   std::ostringstream oss;
-  for (size_t i = 0; i < DigestType::DIGESTSIZE; ++i)
+  for (size_t i = 0; i < DigestType::DIGESTSIZE; ++i) {
     oss << std::setfill('0') << std::setw(2) << std::hex << std::uppercase << (0xff & (unsigned int)d[i]);
```
Comment (arc-vmware) [inline util/src/Digest.cpp:102]: should we use the apis from hextools to print the digest in hexa decimal format.

Comment (pkthapa) [reply util/src/Digest.cpp]: `toString()` replaced with hex tool api in [this commit](https://github.com/vmware/concord-bft/pull/2457/commits/187c7eb432d9e177dce8488a23aa554ead109916).

**Review by arc-vmware — APPROVED** [review]
LGTM.

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp [hunk 2]
```diff
@@ -1055,9 +1055,11 @@ struct VerifyDbCheckpoint {
     os << "  }";
     return os.str();
   }
-  bool isSame(const Digest &d, const STDigest &st) const {
+
+  bool isSame(const Digest &d, const Digest &st) const {
```
Comment (mdarade) [inline kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp:1059]: Is this function really required?
Can existing operator== be used?

File: storage/src/s3/integrity_checker.cpp [hunk 3]
```diff
@@ -23,6 +23,9 @@
 #include "bcstatetransfer/BCStateTran.hpp"
 #include "KeyfileIOUtils.hpp"
 #include "direct_kv_storage_factory.h"
+#include "Digest.hpp"
+
+using bftEngine::impl::Digest;
```
Comment (mdarade) [inline storage/src/s3/integrity_checker.cpp:28]: Again, should namespace be changed?

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 4]
```diff
@@ -45,6 +45,7 @@ using std::chrono::system_clock;
 using namespace std::placeholders;
 using namespace concord::diagnostics;
 using namespace concord::util;
+using bftEngine::impl::Digest;
```
Comment (mdarade) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:48]: Once you modify namespace, I don't think this is required.

File: bftengine/src/bcstatetransfer/DBDataStore.cpp [hunk 5]
```diff
@@ -3,7 +3,6 @@
 #include "storage/db_interface.h"
 #include "Serializable.h"
 
-using bftEngine::bcst::BLOCK_DIGEST_SIZE;
```
Comment (mdarade) [inline bftengine/src/bcstatetransfer/DBDataStore.cpp:6]: Can you remove/replace all occurrences of BLOCK_DIGEST_SIZE if DIGEST_SIZE is having same value?

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp [hunk 6]
```diff
@@ -1099,7 +1101,9 @@ struct VerifyDbCheckpoint {
               const CheckpointDesc &desc,
               bool verifySignature,
               const std::map<ReplicaId, std::unique_ptr<IVerifier>> &verifiers) const {
-    auto is_digest_valid = (isSame(msg.digestOfState(), desc.digestOfLastBlock) && (msg.state() == desc.lastBlock));
+    // auto is_digest_valid = (isSame(msg.digestOfState(), desc.digestOfLastBlock) && (msg.state() == desc.lastBlock));
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp]: Removed.

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp [hunk 7]
```diff
@@ -1021,7 +1021,7 @@ struct GetColumnFamilyStats {
 
 struct VerifyDbCheckpoint {
   using CheckPointMsgStatus = std::vector<std::pair<const CheckpointMsg &, bool>>;
-  using STDigest = bftEngine::bcst::impl::STDigest;
+  using Digest = bftEngine::impl::Digest;
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp]: Done.

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp (hunk 2 above)
Comment (pkthapa) [reply kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp]: `isSame()` removed.

File: storage/src/s3/integrity_checker.cpp (hunk 3 above)
Comment (pkthapa) [reply storage/src/s3/integrity_checker.cpp]: Done.

File: util/include/Digest.hpp [hunk 8]
```diff
@@ -14,7 +14,7 @@
 #include <memory.h>
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply util/include/Digest.hpp:14]: Done.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 4 above)
Comment (pkthapa) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: Done.

File: bftengine/src/bcstatetransfer/DBDataStore.cpp (hunk 5 above)
Comment (pkthapa) [reply bftengine/src/bcstatetransfer/DBDataStore.cpp:6]: Changed.

**Comment by pkthapa** [comment]
> Also, Do you need 4 individual commits or single commit is enough?

Commits reduced to 2 numbers.

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mdarade) [reply kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp]: I don't see it changed for Digest.hpp, Digest.cpp files.

**Comment by mdarade** [comment]
As I general comment, I would suggest to run ASAN.

File: kvbc/include/block_digest.h [hunk 9]
```diff
@@ -9,8 +9,6 @@
 
 namespace concord::kvbc {
```
Comment (mdarade) [inline kvbc/include/block_digest.h:10]: Do we need this file just for single declaration?
I think it can be moved to some other header.

Comment (pkthapa) [reply kvbc/include/block_digest.h:10]: I suggest to keep it as is because it is used only within `kvbc` folder.

**Comment by pkthapa** [comment]
> As I general comment, I would suggest to run ASAN.

ASAN & TSAN are already scheduled for weekly runs.

File: kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pkthapa) [reply kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp]: Addressed in https://github.com/vmware/concord-bft/pull/2457/commits/71b5221b2dbaf0eee972ea40e599cf9059ff05cf

File: storage/src/s3/integrity_checker.cpp [hunk 10]
```diff
@@ -23,6 +23,7 @@
 #include "bcstatetransfer/BCStateTran.hpp"
 #include "KeyfileIOUtils.hpp"
 #include "direct_kv_storage_factory.h"
+#include "Digest.hpp"
```
Comment (mdarade) [inline storage/src/s3/integrity_checker.cpp:26]: If header file doesn't have it included why cpp file required it?

**Review by mdarade — APPROVED** [review]
Looks good.
Thanks Pankaj.

File: storage/src/s3/integrity_checker.cpp (hunk 10 above)
Comment (pkthapa) [reply storage/src/s3/integrity_checker.cpp]: `Digest.hpp` inclusion removed.

File: util/include/Digest.hpp [hunk 11]
```diff
@@ -14,17 +14,21 @@
...
 #include <string>
-#include "DigestType.h"
+#include <array>
+#include "DigestType.hpp"
 
-namespace bftEngine {
-namespace impl {
+namespace concord::util::digest {
+
+using BlockDigest = std::array<std::uint8_t, DIGEST_SIZE>;
```
Comment (toly-kournik) [inline util/include/Digest.hpp:22]: BlockDigest definition should be part of kvbc, utils knows nothing about blocks

**Comment by toly-kournik** [comment]
@pkthapa , we have `concord::util::SHA3_256` defined in sha_hash.hpp.
It uses `openssl/evp.h`. So why would we continue and maintain a Digest class based on crypto++ library, which we are going to get rid off in the nearest future?
Digest can be typedef'ed to it.
Another point is that theoretically STDigest can differ from Digest. So it could retain its name and typedef'ed as well.

File: util/include/Digest.hpp (hunk 11 above)
Comment (pkthapa) [reply util/include/Digest.hpp:22]: Moving this definition to `kvbc/include/categorization/blocks.h`

**Comment by pkthapa** [comment]
> @pkthapa , we have `concord::util::SHA3_256` defined in sha_hash.hpp. It uses `openssl/evp.h`. So why would we continue and maintain a Digest class based on crypto++ library, which we are going to get rid off in the nearest future? Digest can be typedef'ed to it. Another point is that theoretically STDigest can differ from Digest. So it could retain its name and typedef'ed as well.

Yes. I agree that our ultimate aim is to get rid of `Crypto++` library. However, with this approach we introduced a wrapper (`Digest` class) for different libraries (and/or algorithms) which gives us the flexibility to add different libraries for digest generation.

Earlier, `STDigest` and `Digest` classes represented similar features. Hence, I felt no legit usage of retaining another class (`STDigest`) having duplicate features of `Digest` class.

**Pull request merged**